# What's the Opposite of a Maddux?

## Data

Stats LLC started tracking pitch counts in 1988, therefore generations of pitchers could not be used for the following analysis. I've scraped all the game logs for the top 50 pitchers with the most complete games.

In [11]:
import pandas as pd

data = pd.read_csv('csvs/game_logs.csv')
players = pd.read_csv('csvs/player_names.csv', index_col='player_id')

## Check Data

Check to ensure data looks right and contains the correct number of players

In [12]:
print data.loc[0]
print len(data.pivot_table(index='player_id', values=['runs'], aggfunc='sum'))

player_id        abbotji01
team                   CAL
opp                    SEA
inngs                 GS-5
result               L,0-7
pitches                 83
ip                     4.2
runs                     6
year                  1989
month                    4
day                      8
entered       1t start tie
Name: 0, dtype: object
50


## Analysis

### Total number of complete games

In [13]:
complete_games = data[data['inngs'].isin(['SHO', 'CG'])]
shutouts = data[data['inngs'] == 'SHO']

### Average Pitches per Complete Game

In [14]:
complete_games.groupby('player_id').aggregate({'pitches': 'mean', 'inngs': 'count'})\
    .sort_values('pitches', ascending=True).rename(columns={'inngs':'Complete Games'}).join(players)

Complete Games     pitches       player_name
player_id                                              
tewksbo01              29   99.857143     Bob Tewksbury
buehrma01              33  106.606061      Mark Buehrle
maddugr01             107  106.924528       Greg Maddux
radkebr01              37  107.916667        Brad Radke
carpech01              33  108.454545   Chris Carpenter
hallaro01              67  108.582090      Roy Halladay
mulhote01              46  109.395349  Terry Mulholland
leecl02                29  109.551724         Cliff Lee
sabatc.01              38  109.736842     C.C. Sabathia
wellsda01              54  109.981481       David Wells
hershor01              33  111.121212    Orel Hershiser
glavito02              56  111.444444       Tom Glavine
colonba01              36  111.600000     Bartolo Colon
bosioch01              37  111.611111       Chris Bosio
swindgr01              35  112.485714     Greg Swindell
moyerja01              31  112.933333       Jamie Moyer
drabedo01              52  113.555556       Doug Drabek
martide01              50  113.604167   Dennis Martinez
navarja01              32  114.032258     Jaime Navarro
abbotji01              31  114.066667        Jim Abbott
saberbr01              45  114.800000   Bret Saberhagen
hentgpa01              34  114.838710       Pat Hentgen
fernaal01              33  114.857143    Alex Fernandez
mussimi01              57  115.160714      Mike Mussina
brownke01              72  115.200000       Kevin Brown
nagych01               31  115.290323      Charles Nagy
ericksc01              51  115.500000    Scott Erickson
astacpe01              31  115.793103     Pedro Astacio
schilcu01              83  115.887500    Curt Schilling
smoltjo01              53  116.061224       John Smoltz
wakefti01              33  116.419355     Tim Wakefield
belchti01              42  117.000000       Tim Belcher
morrija02              52  117.220000       Jack Morris
hurstbr01              36  117.388889       Bruce Hurst
rogerke01              36  117.617647      Kenny Rogers
martipe02              46  117.906977    Pedro Martinez
appieke01              34  118.529412      Kevin Appier
violafr01              34  118.882353       Frank Viola
mooremi01              32  119.531250        Mike Moore
stewada01              38  119.729730      Dave Stewart
mcdowja01              62  120.758065     Jack McDowell
finlech01              63  120.868852      Chuck Finley
coneda01               55  121.862745        David Cone
hernali01              50  122.080000   Livan Hernandez
candito01              42  123.047619     Tom Candiotti
langsma01              51  123.460000     Mark Langston
wittbo01               46  125.630435        Bobby Witt
johnsra05             100  126.173469     Randy Johnson
clemero02              82  126.341463     Roger Clemens
martira02              37  126.909091    Ramon Martinez

### Average Pitches per Shutout

In [15]:
avg_p_per_sho = shutouts.groupby('player_id').aggregate({'pitches': 'mean', 'inngs': 'count'})\
    .sort_values('pitches', ascending=True).rename(columns={'inngs':'sho', 'pitches': 'p'}).join(players)

print avg_p_per_sho

           sho           p       player_name
player_id                                   
tewksbo01    7   95.000000     Bob Tewksbury
bosioch01    8  100.500000       Chris Bosio
maddugr01   34  102.205882       Greg Maddux
radkebr01   10  105.800000        Brad Radke
moyerja01    9  106.222222       Jamie Moyer
carpech01   15  106.466667   Chris Carpenter
buehrma01   10  106.500000      Mark Buehrle
mulhote01   10  107.000000  Terry Mulholland
hallaro01   20  107.000000      Roy Halladay
colonba01   13  107.230769     Bartolo Colon
nagych01     6  108.500000      Charles Nagy
glavito02   25  108.708333       Tom Glavine
abbotji01    6  109.000000        Jim Abbott
drabedo01   20  109.222222       Doug Drabek
wellsda01   12  109.250000       David Wells
hershor01   14  110.142857    Orel Hershiser
leecl02     12  110.916667         Cliff Lee
morrija02    9  111.222222       Jack Morris
martide01   18  111.388889   Dennis Martinez
sabatc.01   12  111.583333     C.C. Sabathia
candito01 

Bob Tewksbury was the most efficient pitcher during his complete games and shutouts, averaging right around 100 pitches during his complete games and 95 during his shutouts. However, from 1988 to 1998, he threw only 29 complete games. Greg Maddux played for an additional ten years and threw 107 complete games from 1988 to 2008. He averaged right around 107 pitches per complete game.

Randy Johnson, having thrown the second most complete games (100) averaged 126 pitches per game. That's 20 more pitches per game!

### Who has thrown the most Madduxes??

In [16]:
under_100 = shutouts[(shutouts['pitches'] < 100) & (shutouts['ip'] >= 9)]

under_100 = under_100.groupby('player_id').aggregate({'pitches': 'mean', 'inngs': 'count'})\
    .sort_values('inngs', ascending=False).rename(columns={'inngs':'sho', 'pitches': 'p'}).join(players)
print under_100

           sho          p       player_name
player_id                                  
maddugr01   13  91.692308       Greg Maddux
tewksbo01    6  91.500000     Bob Tewksbury
hallaro01    5  95.000000      Roy Halladay
glavito02    5  92.600000       Tom Glavine
moyerja01    4  93.750000       Jamie Moyer
bosioch01    4  88.000000       Chris Bosio
colonba01    4  95.000000     Bartolo Colon
wellsda01    3  94.333333       David Wells
navarja01    3  93.333333     Jaime Navarro
radkebr01    3  95.000000        Brad Radke
schilcu01    3  95.333333    Curt Schilling
mulhote01    3  97.000000  Terry Mulholland
carpech01    3  97.000000   Chris Carpenter
mussimi01    2  96.000000      Mike Mussina
drabedo01    2  86.500000       Doug Drabek
clemero02    2  90.000000     Roger Clemens
hershor01    2  96.500000    Orel Hershiser
rogerke01    2  95.500000      Kenny Rogers
leecl02      2  96.000000         Cliff Lee
buehrma01    2  91.500000      Mark Buehrle
brownke01    2  88.500000       

### Who has thrown the most Johnsons??

In [17]:
over_125 = shutouts[(shutouts['pitches'] > 125) & (shutouts['ip'] >= 9)]
over_125 = over_125.groupby('player_id').aggregate({'pitches': 'mean', 'inngs': 'count'})\
    .sort_values('inngs', ascending=False).rename(columns={'inngs':'sho', 'pitches': 'p'}).join(players)

print over_125

           sho           p      player_name
player_id                                  
clemero02   20  137.400000    Roger Clemens
johnsra05   20  135.600000    Randy Johnson
coneda01    10  135.700000       David Cone
schilcu01    8  131.875000   Curt Schilling
wittbo01     7  132.285714       Bobby Witt
martira02    7  132.000000   Ramon Martinez
langsma01    6  133.666667    Mark Langston
martipe02    5  131.000000   Pedro Martinez
mussimi01    5  132.000000     Mike Mussina
hurstbr01    5  131.800000      Bruce Hurst
brownke01    4  127.750000      Kevin Brown
mcdowja01    4  133.750000    Jack McDowell
finlech01    4  137.250000     Chuck Finley
belchti01    3  131.666667      Tim Belcher
smoltjo01    3  132.666667      John Smoltz
drabedo01    3  132.666667      Doug Drabek
stewada01    3  130.333333     Dave Stewart
glavito02    3  127.666667      Tom Glavine
hernali01    3  133.333333  Livan Hernandez
swindgr01    3  133.666667    Greg Swindell
martide01    2  133.500000  Denn

## Conclusion

Given the available data, a Maddux is the best term for a complete game shutout under 100 pitches. If we were to name a highly inefficient complete game shutout, it'd be called a Johnson, named after Randy Johnson who averaged 126 pitches during his shutouts and never threw a Maddux in his entire career!

## Create json

In [24]:
# avg_p_per_sho
# under_100
# over_125
with open('maddux.json', 'wa') as f:
    f.write('{ "avg_p_per_sho":\n')
    f.write(avg_p_per_sho.to_json(orient='records'))
    f.write(",\n")

    f.write('"over_125":\n')
    f.write(over_125[:10].to_json(orient='records'))
    f.write(",\n")

    f.write('"under_100":\n')
    f.write(under_100[:10].to_json(orient='records'))
    f.write("}")

## Resources

* [https://en.wikipedia.org/wiki/Pitch_count](https://en.wikipedia.org/wiki/Pitch_count)
* [http://baseball-reference.com/](http://baseball-reference.com/)
* [http://www.fangraphs.com/](http://www.fangraphs.com/)